### PR TITLE
Submodules

### PR DESCRIPTION
**Pull request by mitchspano**
Remove static copies of referenced packages in favor of using submodules

### CODE REVIEW

**Review by hmayerhutte — COMMENTED** [review]
oh wow - @mmoyaferrer do I understand this correctly that with submodules, we can just reference repos, and the latest is then used? are those then "resolved" at the moment of cloning the repo? or when building the unlocked package that contains those?

**Comment by hmayerhutte** [comment]
@mitchspano I only noticed after submitting my comment that this contribution is actually by you, not @mmoyaferrer, who made me a reviewer. Thanks a lot for your contribution - I'll catch up on the submodules approach, this def looks like a great improvement in regards to our ability to manage recipes which re-use third party open source 🙌
@mmoyaferrer could you elaborate on potential effect of this on the unlocked package version of the frameworks which we also provide? Any plans to rebuild versions regularly, as the frameworks move ahead? Would it probably make sense to set up a scheduled Github Action for automating that?

**Comment by amtrack** [comment]
@hmayerhutte Renovate seems to support [updating Git submodules](https://docs.renovatebot.com/modules/manager/git-submodules/), maybe Dependabot as well but I couldn't find it quickly.

**Comment by mmoyaferrer** [comment]
> oh wow - @mmoyaferrer do I understand this correctly that with submodules, we can just reference repos, and the latest is then used? are those then "resolved" at the moment of cloning the repo? or when building the unlocked package that contains those?

A Git submodule references to a repository for a specific commit, "snapshot".

From https://www.atlassian.com/git/tutorials/git-submodule:
*Git submodules allow you to keep a Git repository as a subdirectory of another Git repository. Git submodules are simply a reference to another repository at a particular snapshot in time. Git submodules enable a Git repository to incorporate and track version history of external code.*

Therefore when cloning it will clone the linked repository for that specific commit.

**Comment by mmoyaferrer** [comment]
> @mitchspano I only noticed after submitting my comment that this contribution is actually by you, not @mmoyaferrer, who made me a reviewer. Thanks a lot for your contribution - I'll catch up on the submodules approach, this def looks like a great improvement in regards to our ability to manage recipes which re-use third party open source 🙌 @mmoyaferrer could you elaborate on potential effect of this on the unlocked package version of the frameworks which we also provide? Any plans to rebuild versions regularly, as the frameworks move ahead? Would it probably make sense to set up a scheduled Github Action for automating that?

@mitchspano is lightning-speed!

Regarding your questions:
> potential effect of this on the unlocked package version of the frameworks which we also provide

No effect in our current version, as I cannot see any major changes in the frameworks (most have not been changed during recent weeks) hence from my point of view all good to go.

> Any plans to rebuild versions regularly, as the frameworks move ahead? Would it probably make sense to set up a scheduled Github Action for automating that?

Let's talk about this, I don't have a fixed answer since not all frameworks follow Github releases, but we can follow other solutions.

**Review by mmoyaferrer — APPROVED** [review]

**Comment by jongpie** [comment]
@mitchspano @hmayerhutte @amtrack @mmoyaferrer one issue with this approach is it will pull in a lot of extra metadata from Nebula Logger's repo. Nebula Logger's repo is a monorepo, containing metadata for the core package, as well as several plugins, recipes, test metadata, and some deprecated metadata. Prior to this PR, it looks like only the 'core' folder was being used (which is probably all you want). With the submodule approach, I think  _everything_ will get included, including a lot of metadata that definitely should not be packaged.

I'm not sure what options there are to handle this when using submodules, but **you will break some orgs if you package everything from Nebula Logger.**

**Comment by mmoyaferrer** [comment]
> @mitchspano @hmayerhutte @amtrack @mmoyaferrer one issue with this approach is it will pull in a lot of extra metadata from Nebula Logger's repo. Nebula Logger's repo is a monorepo, containing metadata for the core package, as well as several plugins, recipes, test metadata, and some deprecated metadata. Prior to this PR, it looks like only the 'core' folder was being used (which is probably all you want). With the submodule approach, I think _everything_ will get included, including a lot of metadata that definitely should not be packaged.
> 
> I'm not sure what options there are to handle this when using submodules, but **you will break some orgs if you package everything from Nebula Logger.**

Good catch, thank you for the comment. 

@mitchspano I would initially suggest to keep Nebula-core as it is now, and others as submodules. It's a bit heart-breaking because the main benefit of submodules was for Nebula which is the one with the most contribution, but it's the most balanced approach I see. Thoughts?

**Comment by jongpie** [comment]
@mmoyaferrer it looks like [you made a commit](https://github.com/hutte-recipes/hutte-project-template/commit/c5876b8622c4c1adb32cb9c0f4f01dc7ed04834c) to exclude the frameworks from being pushed or pulled... If the metadata in the `frameworks` directory isn't intended be pushed/pulled for this repo, then it feels like all of the frameworks metadata (everything included in your unlocked package) should be a separate repo from this one. Right now, people would clone this repo, and get all of the included framework's metadata (which is a lot of files to clone), even though that metadata would never be deployed or used. It also makes it difficult to use `.forceignore` to help with this submodule situation.

If you split it into a separate repo, then you could have some additional lines in the new repo's `.forceignore` file to exclude the extra metadata in Nebula Logger, which would let you use the submodules approach that @mitchspano is suggesting

```bash
// Excluded extra folders from Nebula Logger
frameworks/nebula-logger/extra-tests
frameworks/nebula-logger/managed-package
frameworks/nebula-logger/plugins
frameworks/nebula-logger/recipes
```

**Comment by mmoyaferrer** [comment]
> @mmoyaferrer it looks like [you made a commit](https://github.com/hutte-recipes/hutte-project-template/commit/c5876b8622c4c1adb32cb9c0f4f01dc7ed04834c) to exclude the frameworks from being pushed or pulled... If the metadata in the `frameworks` directory isn't intended be pushed/pulled for this repo, then it feels like all of the frameworks metadata (everything included in your unlocked package) should be a separate repo from this one. Right now, people would clone this repo, and get all of the included framework's metadata (which is a lot of files to clone), even though that metadata would never be deployed or used. It also makes it difficult to use `.forceignore` to help with this submodule situation.
> 
> If you split it into a separate repo, then you could have some additional lines in the new repo's `.forceignore` file to exclude the extra metadata in Nebula Logger, which would let you use the submodules approach that @mitchspano is suggesting
> 
> ```shell
> // Excluded extra folders from Nebula Logger
> frameworks/nebula-logger/extra-tests
> frameworks/nebula-logger/managed-package
> frameworks/nebula-logger/plugins
> frameworks/nebula-logger/recipes
> ```

The main purpose of having `frameworks` folder in the repository is to provide the developers an easy way to access the frameworks content. For example, if I'm a developer working on a SOQL selector, I can open `SOQL.cls` to understand how it works in a simple and easy way, and the same concept for the other frameworks.

While I understand that the git clone and maintenance can be improved using submodules, and the reason why you would create an independent repository, I would not separate it to an independent repository as I would be breaking the main purpose of the `frameworks`folder.

I would like @hmayerhutte or @amtrack view on the topic.

**Comment by hmayerhutte** [comment]
@mmoyaferrer I see the point that @jongpie makes about .forceignore - it could be misleading for users of this template who do not go with the suggested approach of installing the frameworks from the unlocked package which we provide, but would prefer to push them. While I understand your preference of giving users access to the apex classes directly in our template, I wonder if ease of maintenance wouldn't outweigh that benefit. 
I think having curated frameworks as part of our template is a great benefit (in combination with providing a periodically updated unlocked package containing them) even if it doesn't contain their source.
To sum it up: I sympathize with the submodules approach + separate repo with nebula and forceignore to deal with the extra metadata that shouldn't be pushed as outlined by @jongpie

**Comment by mmoyaferrer** [comment]
> @mmoyaferrer I see the point that @jongpie makes about .forceignore - it could be misleading for users of this template who do not go with the suggested approach of installing the frameworks from the unlocked package which we provide, but would prefer to push them. While I understand your preference of giving users access to the apex classes directly in our template, I wonder if ease of maintenance wouldn't outweigh that benefit. I think having curated frameworks as part of our template is a great benefit (in combination with providing a periodically updated unlocked package containing them) even if it doesn't contain their source. To sum it up: I sympathize with the submodules approach + separate repo with nebula and forceignore to deal with the extra metadata that shouldn't be pushed as outlined by @jongpie

Thanks for your feedback @hmayerhutte , I understand your points. Let's do that :) 
@mitchspano I will decline this repo and follow the suggested approach by @jongpie 🚀 Thanks all for your inputs!
